File: mobile/lib/models/order.dart
class OrderItem {
  final String productId;
  final String name;
  final double price;
  final int quantity;
  final double total;
  final int kotSentQuantity;
  final String imageUrl;

  OrderItem({required this.productId, required this.name, required this.price, required this.quantity, required this.total, this.kotSentQuantity = 0, this.imageUrl = ''});

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    final productField = json['product'];
    return OrderItem(
      productId: productField is Map ? productField['_id'] as String : ((productField as String?) ?? (json['quickItem'] is Map ? 'quick:${json['quickItem']['name']}:${json['quickItem']['price']}' : '')),
      name: json['name'] as String? ?? '',
      price: (json['price'] as num?)?.toDouble() ?? 0,
      quantity: (json['quantity'] as num?)?.toInt() ?? 0,
      total: (json['total'] as num?)?.toDouble() ?? 0,
      kotSentQuantity: (json['kotSentQuantity'] as num?)?.toInt() ?? 0,
      imageUrl: productField is Map ? (productField['imageUrl'] as String? ?? '') : '',
    );
  }
}


class KotTicket {
  final String id;
  final int kotNumber;
  final List<KotItem> items;
  final String status;
  final String? sentBy;
  final DateTime createdAt;

  KotTicket({required this.id, required this.kotNumber, required this.items, required this.status, this.sentBy, required this.createdAt});

  factory KotTicket.fromJson(Map<String, dynamic> json) => KotTicket(
    id: json['_id']?.toString() ?? '',
    kotNumber: (json['kotNumber'] as num?)?.toInt() ?? 0,
    items: (json['items'] as List<dynamic>? ?? []).map((e) => KotItem.fromJson(e as Map<String, dynamic>)).toList(),
    status: json['status']?.toString() ?? 'new',
    sentBy: json['sentBy'] is Map ? json['sentBy']['name']?.toString() : json['sentBy']?.toString(),
    createdAt: DateTime.tryParse(json['createdAt']?.toString() ?? '') ?? DateTime.now(),
  );
}

class KotItem {
  final String name;
  final int quantity;
  KotItem({required this.name, required this.quantity});
  factory KotItem.fromJson(Map<String, dynamic> json) => KotItem(name: json['name']?.toString() ?? '', quantity: (json['quantity'] as num?)?.toInt() ?? 0);
}

class PaymentBreakdown {
  final double cash;
  final double upi;
  final double credit;

  PaymentBreakdown({this.cash = 0, this.upi = 0, this.credit = 0});

  factory PaymentBreakdown.fromJson(Map<String, dynamic>? json) {
    if (json == null) return PaymentBreakdown();
    return PaymentBreakdown(
      cash: (json['cash'] as num?)?.toDouble() ?? 0,
      upi: (json['upi'] as num?)?.toDouble() ?? 0,
      credit: (json['credit'] as num?)?.toDouble() ?? 0,
    );
  }
}

class DeliveryInfo {
  final String? address;
  final String? phone;
  DeliveryInfo({this.address, this.phone});

  factory DeliveryInfo.fromJson(Map<String, dynamic>? json) {
    if (json == null) return DeliveryInfo();
    return DeliveryInfo(address: json['address'] as String?, phone: json['phone'] as String?);
  }
}

/// Contact info a customer optionally gave when placing a QR order directly
/// (no staff/login involved). Absent on regular POS orders.
class QrCustomerContact {
  final String? name;
  final String? phone;
  QrCustomerContact({this.name, this.phone});

  factory QrCustomerContact.fromJson(Map<String, dynamic>? json) {
    if (json == null) return QrCustomerContact();
    return QrCustomerContact(name: json['name'] as String?, phone: json['phone'] as String?);
  }
}

class Order {
  final String id;
  final int orderNumber;
  final List<OrderItem> items;
  final double subtotal;
  final double discount;
  final double loyaltyPointsUsed;
  final double tax;
  final double grandTotal;
  final String orderType; // dine_in | takeaway | delivery
  final String orderSource; // pos | qr — how this order was placed
  final String? tableId;
  final String? tableName;
  final String? tableCustomerLabel;
  final QrCustomerContact? qrCustomerContact;
  final DeliveryInfo? deliveryInfo;
  final String paymentMethod; // CASH | UPI | CREDIT | MIXED
  final String paymentStatus; // pending | payment_initiated | paid | failed | cancelled
  final PaymentBreakdown paymentBreakdown;
  final String? upiReference;
  final double? amountReceived;
  final double changeReturned;
  final String? customerId;
  final String? customerName;
  final String? customerPhone;
  final String? notes;
  final String staffId;
  final String? staffName; // this is the current "attended by" employee
  final String status; // legacy billing/order lifecycle
  final String kitchenStatus; // new | preparing | ready | served
  final List<KotTicket> kots;
  final DateTime createdAt;
  final DateTime? estimatedReadyAt;

  Order({
    required this.id,
    required this.orderNumber,
    required this.items,
    required this.subtotal,
    required this.discount,
    this.loyaltyPointsUsed = 0,
    required this.tax,
    required this.grandTotal,
    this.orderType = 'takeaway',
    this.orderSource = 'pos',
    this.tableId,
    this.tableName,
    this.tableCustomerLabel,
    this.qrCustomerContact,
    this.deliveryInfo,
    required this.paymentMethod,
    this.paymentStatus = 'pending',
    required this.paymentBreakdown,
    this.upiReference,
    this.amountReceived,
    this.changeReturned = 0,
    this.customerId,
    this.customerName,
    this.customerPhone,
    this.notes,
    required this.staffId,
    this.staffName,
    required this.status,
    this.kitchenStatus = 'new',
    this.kots = const [],
    required this.createdAt,
    this.estimatedReadyAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    final customerField = json['customer'];
    final staffField = json['staff'];
    final tableField = json['table'];

    return Order(
      id: json['_id'] as String,
      orderNumber: (json['orderNumber'] as num?)?.toInt() ?? 0,
      items: (json['items'] as List<dynamic>? ?? []).map((e) => OrderItem.fromJson(e as Map<String, dynamic>)).toList(),
      subtotal: (json['subtotal'] as num?)?.toDouble() ?? 0,
      discount: (json['discount'] as num?)?.toDouble() ?? 0,
      loyaltyPointsUsed: (json['loyaltyPointsUsed'] as num?)?.toDouble() ?? 0,
      tax: (json['tax'] as num?)?.toDouble() ?? 0,
      grandTotal: (json['grandTotal'] as num?)?.toDouble() ?? 0,
      orderType: json['orderType'] as String? ?? 'takeaway',
      orderSource: json['orderSource'] as String? ?? 'pos',
      tableId: tableField is Map ? tableField['_id'] as String? : tableField as String?,
      tableName: tableField is Map ? tableField['name'] as String? : null,
      tableCustomerLabel: json['tableCustomerLabel'] as String?,
      qrCustomerContact: json['qrCustomerContact'] != null
          ? QrCustomerContact.fromJson(json['qrCustomerContact'] as Map<String, dynamic>?)
          : null,
      deliveryInfo: json['deliveryInfo'] != null ? DeliveryInfo.fromJson(json['deliveryInfo'] as Map<String, dynamic>?) : null,
      paymentMethod: json['paymentMethod'] as String? ?? 'CASH',
      paymentStatus: json['paymentStatus'] as String? ?? ((json['status'] as String?) == 'completed' ? 'paid' : 'pending'),
      paymentBreakdown: PaymentBreakdown.fromJson(json['paymentBreakdown'] as Map<String, dynamic>?),
      upiReference: json['upiReference'] as String?,
      amountReceived: (json['amountReceived'] as num?)?.toDouble(),
      changeReturned: (json['changeReturned'] as num?)?.toDouble() ?? 0,
      customerId: customerField is Map ? customerField['_id'] as String? : customerField as String?,
      customerName: customerField is Map ? customerField['name'] as String? : null,
      customerPhone: customerField is Map ? customerField['phone'] as String? : null,
      notes: json['notes'] as String?,
      staffId: staffField is Map ? staffField['_id'] as String? ?? '' : staffField as String? ?? '',
      staffName: staffField is Map ? staffField['name'] as String? : null,
      status: json['status'] as String? ?? 'completed',
      kitchenStatus: json['kitchenStatus'] as String? ?? (json['status'] == 'ready' ? 'ready' : json['status'] == 'preparing' ? 'preparing' : 'new'),
      kots: (json['kots'] as List<dynamic>? ?? []).map((e) => KotTicket.fromJson(e as Map<String, dynamic>)).toList(),
      createdAt: DateTime.tryParse(json['createdAt'] as String? ?? '') ?? DateTime.now(),
      estimatedReadyAt: DateTime.tryParse(json['estimatedReadyAt'] as String? ?? ''),
    );
  }

  bool get isQrOrder => orderSource == 'qr';
}

File: backend/src/models/Order.js
const mongoose = require('mongoose');

// OrderItem is embedded as a sub-document of Order. This keeps order creation
// atomic and avoids extra round trips, while still being a clearly defined,
// independently validated schema (satisfies the OrderItem model requirement).
const orderItemSchema = new mongoose.Schema(
  {
    product: { type: mongoose.Schema.Types.ObjectId, ref: 'Product', required: true },
    name: { type: String, required: true }, // snapshot at time of sale
    price: { type: Number, required: true, min: 0 }, // snapshot selling price at time of sale
    quantity: { type: Number, required: true, min: 1 },
    total: { type: Number, required: true, min: 0 },
  },
  { _id: false }
);

const orderSchema = new mongoose.Schema(
  {
    orderNumber: { type: Number, required: true, unique: true }, // sequential, human-readable; unique:true already builds the index

    // Idempotency guard. The Flutter app generates one UUID per checkout
    // attempt (persisted for the lifetime of that bill/payment sheet) and
    // sends it as `clientRequestId`. If a request is retried after a
    // timeout/dropped response — a double-tap the frontend didn't prevent,
    // or a manual "Retry" after a lost response on a Render cold start —
    // the unique index below makes a second insert with the same id fail
    // with a duplicate-key error instead of creating a second order/charge.
    // The controller catches that and returns the original order instead.
    // Sparse so it never applies to legacy/undefined values.
    clientRequestId: { type: String, index: { unique: true, sparse: true } },

    items: { type: [orderItemSchema], required: true, default: [] },

    // Order type: which channel this order came through.
    orderType: { type: String, enum: ['dine_in', 'takeaway', 'delivery'], default: 'takeaway', index: true },
    // Where the order originated. 'qr' orders are placed by a customer via
    // the public /menu page (no logged-in user) and start unattended; 'pos'
    // covers every existing staff-entered order (unchanged default).
    orderSource: { type: String, enum: ['pos', 'qr'], default: 'pos', index: true },
    // Only set for dine_in orders. References the physical Table.
    table: { type: mongoose.Schema.Types.ObjectId, ref: 'Table' },
    // Distinguishes multiple simultaneous customers/orders on the same table
    // (e.g. "Customer 1", "Customer 2"). Purely a display label.
    tableCustomerLabel: { type: String, trim: true },
    // Contact info a customer optionally gives when placing a QR order
    // directly (no account/login). Not used by POS-entered orders.
    qrCustomerContact: {
      name: { type: String, trim: true },
      phone: { type: String, trim: true },
    },
    // Only used for orderType = 'delivery'. Reuses the existing Customer
    // system where possible (see `customer` field below) — these fields
    // exist for delivery-specific details or walk-in delivery without a
    // saved Customer record.
    deliveryInfo: {
      address: { type: String, trim: true },
      phone: { type: String, trim: true },
    },

    subtotal: { type: Number, required: true, min: 0, default: 0 },
    discount: { type: Number, default: 0, min: 0 },
    tax: { type: Number, default: 0, min: 0 },
    grandTotal: { type: Number, required: true, min: 0, default: 0 },

    paymentMethod: { type: String, enum: ['CASH', 'UPI', 'CREDIT', 'MIXED'] },
    // Payment state is deliberately separate from kitchen/order state.
    // QR orders begin pending and become paid only at trusted staff checkout.
    //   pending            - CASH (pay-at-counter) order, or not yet started.
    //   payment_initiated  - customer tapped "Pay with UPI" and was handed
    //                        off to a UPI app. NOT proof of payment — only
    //                        that an attempt started. Set the moment the
    //                        order is created for paymentMethod=UPI.
    //   paid               - verified paid. Only ever set by an authenticated
    //                        staff member at checkout (see orderController),
    //                        after they've confirmed the money actually
    //                        arrived. Never set automatically from the
    //                        public/customer-facing endpoints.
    //   failed             - a payment attempt that is known not to have
    //                        gone through (reserved for future PSP webhook
    //                        integration; unused today).
    //   cancelled          - customer or staff cancelled before payment.
    paymentStatus: {
      type: String,
      enum: ['pending', 'payment_initiated', 'paid', 'failed', 'cancelled'],
      default: 'pending',
      index: true,
    },
    // Set when paymentStatus first becomes 'payment_initiated'. Lets staff
    // see (and, if needed, clean up/cancel) UPI orders where the customer
    // opened a UPI app but never returned to submit a reference.
    paymentInitiatedAt: { type: Date },
    paymentBreakdown: {
      cash: { type: Number, default: 0 },
      upi: { type: Number, default: 0 },
      credit: { type: Number, default: 0 },
    },
    upiReference: { type: String, trim: true },
    amountReceived: { type: Number }, // for cash, amount tendered
    changeReturned: { type: Number, default: 0 },

    customer: { type: mongoose.Schema.Types.ObjectId, ref: 'Customer' }, // required only for CREDIT/MIXED-with-credit, or delivery customer
    notes: { type: String, trim: true },

    // The staff/manager/admin who is handling this order ("attendedBy").
    // Automatically set to the authenticated user at order-start time —
    // never entered manually. Kept as `staff` for backward compatibility
    // with the existing schema/reports; new code should read this as
    // "attendedBy". A QR order has no logged-in user when the customer
    // places it, so it starts with no staff and gets attributed to whoever
    // checks it out (see orderController.checkoutOrder).
    staff: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: function () {
        return this.orderSource !== 'qr';
      },
    },
    // Optional handover trail if a different employee takes over the same
    // customer/order before it's completed. Most orders will have none.
    attendedByHistory: [
      {
        user: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
        from: { type: Date },
        to: { type: Date },
      },
    ],

    status: { type: String, enum: ['open', 'preparing', 'ready', 'completed', 'voided'], default: 'completed' },
    estimatedReadyAt: { type: Date },
    voidedAt: { type: Date },
    voidedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
    voidReason: { type: String, trim: true },
  },
  { timestamps: true }
);

orderSchema.index({ createdAt: -1 });
orderSchema.index({ staff: 1, createdAt: -1 });
orderSchema.index({ customer: 1, createdAt: -1 });
orderSchema.index({ paymentMethod: 1 });
orderSchema.index({ table: 1, status: 1 });
orderSchema.index({ orderType: 1, createdAt: -1 });

module.exports = mongoose.model('Order', orderSchema);
